constrain-lib/src/main/resources/schema/dependency-constraints-schema.json

```diff
@@ -0,0 +1,123 @@
+{
+  "$schema": "https://json-schema.org/draft/2019-09/schema",
+  "type": "object",
+  "title": "Dependency Constraints Schema.",
+  "required": [
+    "dependencyConstraints"
+  ],
+  "properties": {
+    "dependencyConstraints": {
+      "$ref": "#/$defs/dependencyConstraints"
+    }
+  },
+  "additionalProperties": false,
+  "$defs": {
+    "dependencyConstraints": {
+      "type": "array",
+      "title": "Constraints",
+      "description": "A collection of Dependency Constraints.",
+      "items": {
+        "$ref": "#/$defs/dependencyConstraint"
+      }
+    },
+    "dependencyConstraint": {
+      "type": "object",
+      "title": "Constraint",
+      "description": "A single Dependency Constraint.",
+      "required": [
+        "groupId",
+        "name",
+        "suggestedVersion",
```

Comment:
Shouldn't `rejectedVersions` be required and `suggestedVersion` optional instead?

Reply:
+1 on this. In most cases, rejectedVersions will be a better way to express which versions should be avoided, if it is a closed (or half-closed) interval. Usually, the best strategy for automated dependency constraints from a security perspective is, imho, to update to the minimum fixed version available. This way, the risk of breaking changes is reduced, and chances are that the new version will be applied faster if the risk of breaking changes is reduced. If the developer prefers to update to a higher version, I would consider this as out-of-scope for dependency constraints.

Also, it might be possible that at the time a constraint is added by a tool (e.g. Dependabot), there is no fixed version recommendation immediately available (rejectedVersions interval half-open or open, ending with `)`). In this case, the tool might prefer simply indicating that the rejectedVersions should be avoided. This might lead to an unresolvable constraint situation, and this should be carefully considered.

I think it might be worth it to allow suggestedVersion in the specification, but we should note that most of the time this field is probably not really relevant.

Reply:
> Usually, the best strategy for automated dependency constraints from a security perspective is, imho, to update to the minimum fixed version available. This way, the risk of breaking changes is reduced, and chances are that the new version will be applied faster if the risk of breaking changes is reduced.

I agree, and I think this needs to be communicated in the specification.

> I think it might be worth it to allow suggestedVersion in the specification, but we should note that most of the time this field is probably not really relevant.
> Shouldn't `rejectedVersions` be required and `suggestedVersion` optional instead?

Unfortunately no, you need to give Gradle _something_ to update to. That _something_ must be a concrete version that Gradle can resolve. If you don't provide a concrete version for Gradle to resolve to, the Dependency Engine will always fail. We want to increase the likelyhood that a PR generated by Dependabot or similar will be guaranteed to resolve to _some_ version instead of outright failing. Gradle won't automatically go out looking for new versions unless, as a user, you specify that you are willing to accept a range of versions. In a majority of cases, (read: pretty much all active uses of Gradle), this is not the case. If we want to have this file be useful, we need to work within the current status-quo, which is where most users aren't declaring their version ranges as being open currently.

Fundamentally, this feature basically just generates [Gradle Dependency Constraints](https://docs.gradle.org/current/userguide/dependency_constraints.html). So whatever you can do with Dependency Constraints today, this feature can tap into.

Reply:
> Also, it might be possible that at the time a constraint is added by a tool (e.g. Dependabot), there is no fixed version recommendation immediately available (rejectedVersions interval half-open or open, ending with `)`).

I think this is of dubious value to actually support as such a change to add a constraint like this would break the build most likely. As such, forbidding this in the spec seems like a good way to just outright prevent this from being generated. If you encounter this situation, the best thing to happen is for Dependabot to alert users about the vulnerability, but without the ability to generate a PR.

Reply:
Knowing this, I really agree that the rejectedVersions should be optional and suggestedVersion mandatory as you first suggested.

Reply:
This reasoning should be well documented, but given the constraints it makes sense to do it this way.